### ISSUE TITLE
build_kernel.sh fails to integrate qat.conf into .confi

### ISSUE THREAD

**Issue by eadamsintel** [kata-containers/packaging#891]
# Description of problem

The QAT instructions have a section to [Build the Kata Kernel](https://github.com/kata-containers/documentation/blob/master/use-cases/using-Intel-QAT-and-kata.md#download-kata-kernel-source) that integrates some `CRYPTO` options into the kernel to work with an Intel(r) QuickAssist card. The scripts fails to integrate the conf and won't build a .config file. This worked with the previous kernel. 

Below are the crypto options used
```
mkdir -p $GOPATH
cd $GOPATH
go get -v github.com/kata-containers/packaging
cat << EOF > $GOPATH/src/github.com/kata-containers/packaging/kernel/configs/fragments/common/qat.conf
CONFIG_PCIEAER=y
CONFIG_UIO=y
CONFIG_CRYPTO_HW=y
CONFIG_CRYPTO_DEV_QAT_C62XVF=m
CONFIG_CRYPTO_CBC=y
EOF
$GOPATH/src/github.com/kata-containers/packaging/kernel/build-kernel.sh setup
```

# Expected result

The .config file is generated and can be used by `build_kernel.sh` script to create a Kata kernel.

# Actual result

```
INFO: Constructing config from fragments: /home/eadams/src1_7/go/src/github.com/kata-containers/packaging/kernel/configs/fragments/x86_64/.config
INFO: Some CONFIG elements failed to make the final .config:
INFO: Value requested for CONFIG_CRYPTO_DEV_QAT_C62XVF not in final .config
INFO: Generated config file can be found in /home/eadams/src1_7/go/src/github.com/kata-containers/packaging/kernel/configs/fragments/x86_64/.config
ERROR: Failed to construct requested .config file
ERROR: failed to find default config
```

**Comment by grahamwhaley** [comment]
Nominally I suspect that is a 'bug' with the QAT instructions, and they have gotten out of date (rotted) compared with the kernel version now in use.
What the error in the log is telling you is that a `CONFIG` value was requested, that did not make it into the final `.config` file, and that is considered an error you might want to know about.
The fast fix is work out what has changed in the kernel `CONFIG` options, and update the instructions.
The other option is to add the now missing `CONFIG` option to the [whitelist file](https://github.com/kata-containers/packaging/blob/master/kernel/configs/fragments/whitelist.conf), which afaict will then ignore that as an error. But, you would still want to look at the new kernel config setup, as having a var missing that used to be there hints that something has chanced/moved/deprecated, and you'd still want to work out what that was...

/cc @mythi

**Comment by mythi** [comment]
> The fast fix is work out what has changed in the kernel CONFIG options, and update the instructions.

What makes the script to give that error, i.e., what would be the reasons why the config is left out?

**Comment by grahamwhaley** [comment]
> > The fast fix is work out what has changed in the kernel CONFIG options, and update the instructions.
> 
> What makes the script to give that error, i.e., what would be the reasons why the config is left out?

I went and checked the upstream kernel v4.19.86 that is the [current default for Kata](https://github.com/kata-containers/runtime/blob/master/versions.yaml#L158) - looking for the `CONFIG_CRYPTO_DEV_QAT_C62XVF ` - we find that in `drivers/crypto/qat/Kconfig`, and that file has not changed since July 2016 !!! - so, this feels like something else has changed.
@eadamsintel - which kernel or version/branch of Kata were you building against?
And, can you look in the generated `.config` to see what the `QAT` configs do look like?

*tl;dr*
The script uses the [kernel config merge script](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/scripts/kconfig/merge_config.sh) from the upstream Linux kernel to merge `CONFIG` fragments together. If a `CONFIG` item that is requested no longer exists in the kernel version being used, then it gets dropped from the generated `.config` file, and the script will error out, noting that something you asked for didn't make it into the kernel...

**Comment by mythi** [comment]
`CONFIG_CRYPTO_DEV_QAT_C62XVF` has quite a few dependencies that should also be set. Perhaps one of those fails and we only see the error via `CONFIG_CRYPTO_DEV_QAT_C62XVF`.

IIRC I had similar issues with the stackprotector settings: https://github.com/kata-containers/packaging/pull/698

(I never figured out what it means when a `CONF` entry is added without `=m` or `=y`.)

**Comment by eadamsintel** [comment]
The kernel version I successfully built a couple months ago is 4.19.71-49.  The current one is 4.19.86-60 that the .config fails to build from.

**Comment by grahamwhaley** [comment]
OK, progress....
The reason the script fails is that you have requested:
```
CONFIG_CRYPTO_DEV_QAT_C62XVF=m
```

but what you get in the `.config` is
```
CONFIG_CRYPTO_DEV_QAT_C62XVF=y
```

This is because we do not set `CONFIG_MODULES=y` in the kata kernel configs, as we don't normally build modules ;-) That is probably the key difference from last time you tried this I suspect (we may have taken default `MODULES` support out...).

If you then add `CONFIG_MODULES=y` to your list of QAT config fragments.... it gets past your `C62XVF` failure, but then fails with:

```
INFO: Value requested for CONFIG_CRYPTO_FIPS not in final .config
```

and that is because..... `FIPS` only works if modules are signed:

```
        depends on (MODULE_SIG || !MODULES)
```

We added `FIPS` support quite recently, so that is also a new change you'll run into.
/cc @amshinde for FIPS reference.

The quick fix is to add:

```
CONFIG_MODULES=y
CONFIG_MODULE_SIG=y
```

to your QAT config fragments, and that gets you a valid `.config` built. Long term, you'll have to contemplate if that is the best path to take.

> Side note: I notice the QAT config frags are going into the `common` area of the fragments - which, I think is half right - but, I suspect the setting for `C62` is x86 specific? Just a nitpick ;-) If we were to add this to the fragments permanently then I think we'd need to split out the QAT generic bits (that work with PCI plugin cards I suspect for any arch), and any x86 specifics (such as x86 SoC enablement).

**Comment by grahamwhaley** [comment]
Oh, and I'm then not quite sure that your newly signed module will load in your newly built kernel with module sign checks enabled.... it should, but I'm pretty sure we've never tried that with kata before (but, it is a security feature....).
If it fails to load, take out the `MODULE_SIG` *and* `FIPS`, and try again. And then we'll have to sort out what the best right longterm solution is.

**Comment by mythi** [comment]
> If it fails to load, take out the MODULE_SIG and FIPS, and try again.

IIRC the QAT OOT driver installation knows how to sign the modules getting all that documented in Eric's doc adds work.

Is there a way to disable FIPS using fragment overrides?

**Comment by grahamwhaley** [comment]
I don't think you can disable FIPS in the qat fragment by setting it to `=n` - that does disable it in the final .config (probably just due to the order either the crypto or qat fragments are assessed...), but then the fragment generator complains because it sees it was set to `=y` in a fragment, and that is not what turned up in the final .config...

From the kata pov, my gut is telling me we should probably enable module signing if we enable modules, and we should probably write a document section in the generic kata docs discussing kata and kernel modules - as I suspect some folks are either building their own out of tree modules or even loading some binary out of tree modules, and enabling signing might break some use cases...

**Comment by eadamsintel** [comment]
If I try `CONFIG_MODULES=y` and comment out `CONFIG_CRYPTO_FIPS` the `.config` still won't build. I have to also add `CONFIG_MODULE_SIG=y` so that the `.config` builds. However, when I try to build the kernel I get the following errors.

```
...
  HOSTCC  scripts/extract-cert
  HOSTCC  scripts/sign-file
  HOSTCC  scripts/mod/mk_elfconfig
  CC      scripts/mod/empty.o
  CC      scripts/mod/devicetable-offsets.s
scripts/extract-cert.c:21:10: fatal error: openssl/bio.h: No such file or directory
   21 | #include <openssl/bio.h>
      |          ^~~~~~~~~~~~~~~
compilation terminated.
make[1]: *** [scripts/Makefile.host:90: scripts/extract-cert] Error 1
make[1]: *** Waiting for unfinished jobs....
scripts/sign-file.c:25:10: fatal error: openssl/opensslv.h: No such file or directory
   25 | #include <openssl/opensslv.h>
      |          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[1]: *** [scripts/Makefile.host:90: scripts/sign-file] Error 1
  MKELF   scripts/mod/elfconfig.h
  HOSTCC  scripts/mod/modpost.o
  HOSTCC  scripts/mod/sumversion.o
  UPD     scripts/mod/devicetable-offsets.h
  HOSTCC  scripts/mod/file2alias.o
  HOSTLD  arch/x86/tools/relocs
  LD       /home/eadams/src/go/kata-linux-4.19.86-61/tools/objtool/arch/x86/objtool-in.o
  LD       /home/eadams/src/go/kata-linux-4.19.86-61/tools/objtool/libsubcmd-in.o
  AR       /home/eadams/src/go/kata-linux-4.19.86-61/tools/objtool/libsubcmd.a
  HOSTLD  scripts/mod/modpost
make: *** [Makefile:1067: scripts] Error 2
make: *** Waiting for unfinished jobs....
  LD       /home/eadams/src/go/kata-linux-4.19.86-61/tools/objtool/objtool-in.o
  LINK     /home/eadams/src/go/kata-linux-4.19.86-61/tools/objtool/objtool
```

**Comment by amshinde** [comment]
I am not sure why you still need `CONFIG_MODULE_SIG=y` after disabling FIPS.
But for the error you are seeing, you need to install libssl-dev or openssl-devel  depening on your distro.

**Comment by eadamsintel** [comment]
I can't explain it either. The `.config` won't build without it. Thanks for the tip on openssl. I need to update my instructions to make sure 'devpkg-openssl' bundle is added.  I guess things moved around on clear some more.

**Comment by eadamsintel** [comment]
OK I was able to successfully build a 4.19.86 kernel with QAT modules properly loading. The `adf_ctl reset` works fine inside the Kata container. Here is what I had to do.

1. Add in `CONFIG_MODULES=y` and `CONFIG_MODULE_SIG=y` to `qat.conf` in the fragments directory.  
2. Comment out `CONFIG_CRYPTO_FIPS` from `crypto.conf`
3. Add in the `devpkg-openssl` bundle to Clear Linux so it could build.

**Comment by amshinde** [comment]
@eadamsintel If you build with `CONFIG_MODULES` and `CONFIG_MODULE_SIG`, how was `CONFIG_CRYPTO_FIPS` interfering ?

**Comment by eadamsintel** [comment]
@amshinde Actually it wasn't. I just tested and leaving it in does work. I must have misunderstood what Graham was saying. Well, it works with Kata 1.8.2.  I had some issues with the built in kata-runtime 1.9.1 and am not sure why yet.  I am going to try again later with the latest kata in /opt.

So I just need to do 1 and 3 above.

**Comment by eadamsintel** [comment]
OK Well I spoke too quickly. I was using Kata 1.8.2 and everything worked but when I used Kata 1.9.3 and rebuilt a new rootfs with a kata-agent 1.9.3 I had `dmesg` errors for loading out of tree modules.  I guess with these new kernel options I need to sign the module.

```
[    0.281720] intel_qat: loading out-of-tree module taints kernel.
[    0.281767] intel_qat: module verification failed: signature and/or required key missing - tainting kernel
```

**Comment by amshinde** [comment]
@eadamsintel The error message says `tainting kernel`, so the module might still work afaik.
Its just marked the kernel and the module as tainted.
@grahamwhaley can comment on this.

Regarding your second concern, looks like some kernel depends on `CONFIG_MODULES_SIG`.
Can you check for the value of `CONFIG_MODULE_SIG_FORCE` and `CONFIG_MODULE_SIG_ALL` or any other `CONFIG_MODULE_SIG_*` value and make sure it is disabled as well.

As long as `CONFIG_MODULE_SIG_FORCE` is not set to true, you should still be able to work with a tainted kernel I think.

**Comment by eadamsintel** [comment]
OK I am getting closer to isolating my real issue. I actually had gotten the latest QAT driver working with Kata 1.8.2 with my custom kernel/rootfs. The `adf_ctl restart` command did work. However, with Kata 1.9.3 I get the error further below. What happened is that a slightly wrong Docker command line was stuck in my history and I was not passing in the `--device=/dev/vfio/90` and `-v /dev:/dev` when I tested with 1.8.2. That is why I thought it was failing with 1.8.2 even though I previously reported that it worked. It is hard to keep all these variables straight.

What I did figure out is that even with the default kata kernel/image on Kata 1.9.3 if you try to pass in `-v  /dev:/dev` you get the error below.

```
$ docker run -it --runtime=kata-qemu --cap-add=IPC_LOCK --cap-add=SYS_ADMIN --device=/dev/vfio/90 -v /dev:/dev -v ${QAT_CONF_LOCATION}:/etc openssl-qat-engine bash
docker: Error response from daemon: OCI runtime create failed: rpc error: code = Unknown desc = open /run/kata-containers/sandbox/resolv.conf: no such file or directory: unknown.
```

If I do a simple docker run using the default kata kernel/rootfs for 1.9.3 I see the same thing.

```
$ docker run -it --runtime=kata-qemu -v /dev:/dev clearlinux bash
docker: Error response from daemon: OCI runtime create failed: rpc error: code = Unknown desc = open /run/kata-containers/sandbox/resolv.conf: no such file or directory: unknown.
```

Here it is with 1.8.2 using the default kata kernel/rootfs.  Is the issue above a bug or a security feature?  

```
eadams@otc-tme-wf-02~ $ docker run -it --runtime=kata-qemu -v /dev:/dev clearlinux bash
root@a933e562d54a/ #
```

**Comment by grahamwhaley** [comment]
OOI, looking at a `git log --oneline 1.9.0-alpha2..1.9.0-rc0 --no-merges` on the `agent` shows now too much activity (8 commits), but on the `runtime` there seem to be quite a lot of changes (58 commits). Those are the two kata components I would most likely suspect - but, better left to some folks who understand the device handling to look at....

**Comment by eadamsintel** [comment]
@amshinde Is there a more appropriate way to do device pass through and not pass `-v /dev:/dev` as a parameter?   If you run the following command it works on 1.8.2 and older kata-runtime but fails on `1.9.0-rc0` and newer using the default kata kernel/rootfs.  I guess that there is some new security feature added to Kata preventing full access to /dev. Running with `--privileged` still fails. 

```
docker run -it --runtime=kata-runtime -v /dev:/dev clearlinux bash`
docker: Error response from daemon: OCI runtime create failed: Failed to check if grpc server is working: rpc error: code = Unavailable desc = transport is closing: unknown.
```

Both the GPU pass through guide and QAT guide use this `-v /dev:dev` nomenclature.

**Comment by eadamsintel** [comment]
@devimc Has figured out which patch is needed to fix the `-v /dev:/dev` issue.  If the new kata kernel 5.4.x is going to be released soon, I would prefer to wait a bit to get updated instructions that work with that kernel.

**Comment by amshinde** [comment]
@eadamsintel We have switched to using kernel 5.4 kernel in Kata.

**Comment by eadamsintel** [comment]
I will work on this soon as I have another task I am ramping on. I did get my system up and running again with the latest 5.5 host kernel with QAT driver installed and working with my old config.  I was hoping to use `kata-deploy` with the latest 1.10 but it is still in alpha. I was hoping to get the latest 1.11 but it is still at alpha on Docker hub.

**Comment by eadamsintel** [comment]
I confirmed one time on my test system using `kata-deploy:1.11.0-alpha0` and using a kata 5.4.15 kernel with the latest QAT driver that this did work when I used the following options. I was able to `adf_ctl reset` in the container without failure.   It might be some time before Clear Linux is updated to Kata 1.11 with kernel 5.4, but I can get the instructions out before then.

```
$ mkdir -p $GOPATH
$ cd $GOPATH
$ go get -v github.com/kata-containers/packaging
$ cat << EOF > $GOPATH/src/github.com/kata-containers/packaging/kernel/configs/fragments/common/qat.conf
CONFIG_PCIEAER=y
CONFIG_UIO=y
CONFIG_CRYPTO_HW=y
CONFIG_CRYPTO_DEV_QAT_C62XVF=m
CONFIG_CRYPTO_CBC=y
CONFIG_MODULES=y
CONFIG_MODULE_SIG=y
CONFIG_CRYPTO_AUTHENC=y
CONFIG_CRYPTO_DH=yy
EOF
$ $GOPATH/src/github.com/kata-containers/packaging/kernel/build-kernel.sh setup
```
I am having another system setup so I can start from a clean slate and verify the rest of the instructions are still valid.
